frontend/lib/rental-history.tsx

```diff
@@ -0,0 +1,226 @@
+import { SessionUpdatingFormSubmitter } from './session-updating-form-submitter';
+import { RhFormMutation, BlankRhFormInput } from './queries/RhFormMutation';
+import { exactSubsetOrDefault, assertNotNull } from './util';
+import { NextButton, BackButton, CenteredPrimaryButtonLink } from './buttons';
+import { PhoneNumberFormField } from './phone-number-form-field';
+import { AppContext, AppContextType} from './app-context';
+import { Link, Route } from 'react-router-dom';
+import { RhFormInput } from './queries/globalTypes';
+import { RhSendEmailMutation } from './queries/RhSendEmailMutation';
+import * as rhEmailText from '../../common-data/rh.json';
+import { AddressAndBoroughField } from './address-and-borough-form-field';
+import { ConfirmAddressModal, redirectToAddressConfirmationOrNextStep } from './address-confirmation';
+import { getBoroughChoiceLabels, BoroughChoice } from '../../common-data/borough-choices';
+import { ClearSessionButton } from './clear-session-button';
+import { OutboundLink } from './google-analytics';
+import { CustomerSupportLink } from './customer-support-link';
+
+const RH_ICON = "frontend/img/ddo/rent.svg";
+
+/* Rent history welcome page */
+function RentalHistoryWelcome(): JSX.Element {
+    return (
+      <Page title="Request your rent history">
+         <section className="hero is-light">
+          <div className="hero-body">
+            <div className="has-text-centered">
+              <div className="is-inline-block jf-rh-icon">
+                <StaticImage ratio="is-square" src={RH_ICON} alt="" />
+              </div>
+              <h1 className="title is-spaced">
+                Request your rent history in two simple steps!
+              </h1>
+              <p className="subtitle">
+                Let's help you request your <b>rent history</b>! This document, kept by the NY Division of Housing &amp; Community Renewal (DHCR), helps you find out if you're being overcharged.
+              </p>
+              <p className="subtitle is-italic">
+                This service is free, secure, and confidential.
+              </p>
+              <CenteredPrimaryButtonLink to={Routes.locale.rh.form} className="is-large">
+                Start my request
+              </CenteredPrimaryButtonLink>
+            </div>
+          </div>
+        </section>
+      </Page>
+    );
+  }
+
+/** A confirmation modal for when a user's inputted address isn't verified by our geocoder */
+
+function FormConfirmAddressModal(props: { toStep2: string }): JSX.Element {
+  const addrInfo = useContext(AppContext).session.rentalHistoryInfo || BlankRhFormInput;
+  return <ConfirmAddressModal nextStep={props.toStep2} {...addrInfo} />
+}
+
+/** Generates a RhFormInput object based on any data that a logged-in user may have already inputted
+ * If there is no pre-existing user data, we return a blank RhFormInput object 
+ */
+
+function GenerateUserRhFormInput(appContext: AppContextType): RhFormInput {
```

Comment:
> Tried refactoring out the generation of the User's pre-existing form input, but the resulting function still get's caught with a cognitive complexity of 9

Ah interesting that all the cognitive complexity is in this function!  I think it's fine to leave as-is, because as you said, it _is_ fairly readable as-is, though its cognitive complexity does make the case for this being a great candidate for tests.  I think it's ok to not test it right now in favor of merging this PR sooner rather than later, but I don't want to block on it, so we can just file an issue to add a test post-merge.

Reply:
just filed an issue! #895